src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -4,108 +4,104 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
 using System.Formats.Cbor;
-using System.Runtime.Versioning;
 
 namespace System.Security.Cryptography.Cose
 {
-    public sealed class CoseHeaderMap : IEnumerable<(CoseHeaderLabel Label, ReadOnlyMemory<byte> EncodedValue)>
+    public sealed class CoseHeaderMap : IDictionary<CoseHeaderLabel, CoseHeaderValue>, IReadOnlyDictionary<CoseHeaderLabel, CoseHeaderValue>
     {
         private static readonly CoseHeaderMap s_emptyMap = new CoseHeaderMap(isReadOnly: true);
 
         public bool IsReadOnly { get; internal set; }
 
-        private readonly Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>> _headerParameters = new Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>>();
+        public ICollection<CoseHeaderLabel> Keys => _headerParameters.Keys;
 
-        public CoseHeaderMap() : this (isReadOnly: false) { }
+        public ICollection<CoseHeaderValue> Values => _headerParameters.Values;
 
-        internal CoseHeaderMap(bool isReadOnly)
-        {
-            IsReadOnly = isReadOnly;
-        }
+        public int Count => _headerParameters.Count;
 
-        public bool TryGetEncodedValue(CoseHeaderLabel label, out ReadOnlyMemory<byte> encodedValue)
-            => _headerParameters.TryGetValue(label, out encodedValue);
+        IEnumerable<CoseHeaderLabel> IReadOnlyDictionary<CoseHeaderLabel, CoseHeaderValue>.Keys => _headerParameters.Keys;
```

Comment:
Why is there so much code before the ctor?

* Type declaration
  * Consts and static fields
  * Instance fields (and auto-properties, since they're also fields)
  * ctors
  * Everything else

Reply:
The current code aligns with what you just described. 
```cs
private static readonly CoseHeaderMap s_emptyMap = new CoseHeaderMap(isReadOnly: true);
private readonly Dictionary<CoseHeaderLabel, CoseHeaderValue> _headerParameters = new Dictionary<CoseHeaderLabel, CoseHeaderValue>();
private ICollection<KeyValuePair<CoseHeaderLabel, CoseHeaderValue>> HeaderParametersAsCollection => _headerParameters;
public bool IsReadOnly { get; internal set; }
public ICollection<CoseHeaderLabel> Keys => _headerParameters.Keys;
public ICollection<CoseHeaderValue> Values => _headerParameters.Values;
public int Count => _headerParameters.Count;
IEnumerable<CoseHeaderLabel> IReadOnlyDictionary<CoseHeaderLabel, CoseHeaderValue>.Keys => _headerParameters.Keys;
IEnumerable<CoseHeaderValue> IReadOnlyDictionary<CoseHeaderLabel, CoseHeaderValue>.Values => _headerParameters.Values;
public CoseHeaderValue this[CoseHeaderLabel key]
{
    get => _headerParameters[key];
    set
    {
        ValidateIsReadOnly();
        ValidateInsertion(key, value);
        _headerParameters[key] = value;
    }
}

// the ctors are here.
```

Reply:
Oh, are you asking why ctors are _after_ consts, statics, and fields/properties? Not a particular reason, I really didn't thought about it.

Reply:
The `HeaderParametersAsCollection`, `Keys`, `Values`, `Count`, `IReadOnlyDictionary.Keys`, `IReadOnlyDictionary.Values` and indexer should all be after the constructor, as they aren't fields or auto-properties.

```C#
public sealed class CoseHeaderMap
{
    private static readonly CoseHeaderMap s_emptyMap = new CoseHeaderMap(isReadOnly: true);

    private readonly Dictionary<CoseHeaderLabel, CoseHeaderValue> _headerParameters = new Dictionary<CoseHeaderLabel, CoseHeaderValue>();

    public bool IsReadOnly { get; internal set; }

    public CoseHeaderMap() : this(false)
    {
    }

    private CoseHeaderMap(bool isReadOnly)
    {
        IsReadOnly = isReadOnly;
    }

    // other code
}
```